// qbo/src/supportedActions.js
import {
    isIntegerString,
    trim,
    isObject,
    isTrueOrFalse,
    isNonBlankString,
    parseIntString,
    isZeroOrOne,
    cdata
} from "./utilities";

export const supportedActions = {
    API_CopyMasterDetail: {
        destrid: {defaultValue: 0, validate: isIntegerString, normalize: parseIntString},
        sourcerid: {defaultValue: 0, validate: isIntegerString, normalize: parseIntString},
        copyfid: {defaultValue: null, validate: isIntegerString, normalize: parseIntString},
        recurse: {defaultValue: 'true', validate: isTrueOrFalse, normalize: trim},
        relfids: {defaultValue: 'all', validate: isNonBlankString, normalize: trim},
        udata: {defaultValue: null, validate: isNonBlankString, normalize: trim}
    },
    API_DoQuery: {
        query: {defaultValue: null, validate: isNonBlankString, normalize: trim},
        qid: {defaultValue: null, validate: isIntegerString, normalize: parseIntString},
        qname: {defaultValue: null, validate: isNonBlankString, normalize: trim},
        clist: {defaultValue: null, validate: isNonBlankString, normalize: trim},
        slist: {defaultValue: null, validate: isNonBlankString, normalize: trim},
        fmt: {defaultValue: null, validate: isNonBlankString, normalize: trim},
        returnpercentage: {defaultValue: null, validate: isZeroOrOne, normalize: parseIntString},
        options: {defaultValue: null, validate: isNonBlankString, normalize: trim},
        includeRids: {defaultValue: null, validate: isZeroOrOne, normalize: parseIntString},
        udata: {defaultValue: null, validate: isNonBlankString, normalize: trim}
    },
    API_DoQueryCount: {
        query: {defaultValue: null, validate: isNonBlankString, normalize: trim}
    },
    API_GenResultsTable: {
        query: {defaultValue: null, validate: isNonBlankString, normalize: trim},
        qid: {defaultValue: null, validate: isIntegerString, normalize: parseIntString},
        qname: {defaultValue: null, validate: isNonBlankString, normalize: trim},
        clist: {defaultValue: null, validate: isNonBlankString, normalize: trim},
        slist: {defaultValue: null, validate: isNonBlankString, normalize: trim},
        options: {defaultValue: null, validate: isNonBlankString, normalize: trim},
        udata: {defaultValue: null, validate: isNonBlankString, normalize: trim},
        jht: {defaultValue: null},
        jsa: {defaultValue: null, validate: isZeroOrOne, normalize: parseIntString}
    },
    API_GetUserInfo: {
        email: {defaultValue: null, validate: isNonBlankString, normalize: trim},
        udata: {defaultValue: null, validate: isNonBlankString, normalize: trim}
    },
    API_ImportFromCSV: {
        records_csv: {defaultValue: null, validate: isNonBlankString, normalize: cdata},
        clist: {defaultValue: null, validate: isNonBlankString, normalize: trim},
        clist_output: {defaultValue: null, validate: isNonBlankString, normalize: trim},
        skipfirst: {defaultValue: null, validate: isZeroOrOne, normalize: parseIntString},
        msInUTC: {defaultValue: null, validate: isZeroOrOne, normalize: parseIntString},
        mergeFieldId: {defaultValue: null, validate: isNonBlankString, normalize: trim},
        udata: {defaultValue: null, validate: isNonBlankString, normalize: trim}
    },
    API_PurgeRecords: {
        query: {defaultValue: null, validate: isNonBlankString, normalize: trim},
        qid: {defaultValue: null, validate: isIntegerString, normalize: parseIntString},
        qname: {defaultValue: null, validate: isNonBlankString, normalize: trim}
    },
    API_RunImport: {
        "id": {defaultValue: null, validate: isIntegerString, normalize: parseIntString},
        "udata": {defaultValue: null, validate: isNonBlankString, normalize: trim}
    },
    API_GenAddRecordForm: {
        fields: {defaultValue: null, validate: isObject}
    },
    API_GetSchema: {
        "udata": {defaultValue: null, validate: isNonBlankString, normalize: trim}
    },
    API_UserRoles: {
        "udata": {defaultValue: null, validate: isNonBlankString, normalize: trim}
    },
    API_AddUserToRole: {
        "userid": {defaultValue: null, validate: isNonBlankString, normalize: trim},
        "roleid": {defaultValue: null, validate: isNonBlankString, normalize: trim},
        "udata": {defaultValue: null, validate: isNonBlankString, normalize: trim}
    },
    API_GetUserRole: {
        "userid": {defaultValue: null, validate: isNonBlankString, normalize: trim},
        "inclgrps": {defaultValue: null, validate: isNonBlankString, normalize: trim},
        "udata": {defaultValue: null, validate: isNonBlankString, normalize: trim}
    },
    API_ChangeUserRole: {
        "userid": {defaultValue: null, validate: isNonBlankString, normalize: trim},
        "roleid": {defaultValue: null, validate: isNonBlankString, normalize: trim},
        "newroleid": {defaultValue: null, validate: isNonBlankString, normalize: trim},
        "udata": {defaultValue: null, validate: isNonBlankString, normalize: trim}
    },
    API_ProvisionUser: {
        "email": {defaultValue: null, validate: isNonBlankString, normalize: trim},
        "roleid": {defaultValue: null, validate: isNonBlankString, normalize: trim},
        "fname": {defaultValue: null, validate: isNonBlankString, normalize: trim},
        "lname": {defaultValue: null, validate: isNonBlankString, normalize: trim},
        "udata": {defaultValue: null, validate: isNonBlankString, normalize: trim}
    },
    API_UploadFile: {
        "field": {defaultValue: null, validate: isObject},
        "rid": {defaultValue: null, validate: isNonBlankString, normalize: trim}
    },
    API_AddRecord: {
        "fields": {defaultValue: null, validate: isObject},
        "update_id": {defaultValue: null, validate: isNonBlankString, normalize: trim},
        "msInUTC": {defaultValue: null, validate: isZeroOrOne, normalize: parseIntString},
        "udata": {defaultValue: null, validate: isNonBlankString, normalize: trim}
    },
    API_EditRecord: {
        "fields": {defaultValue: null, validate: isObject},
        "update_id": {defaultValue: null, validate: isNonBlankString, normalize: trim},
        "msInUTC": {defaultValue: null, validate: isZeroOrOne, normalize: parseIntString},
        "udata": {defaultValue: null, validate: isNonBlankString, normalize: trim},
        "rid": {defaultValue: null, validate: isNonBlankString, normalize: trim},
        "key": {defaultValue: null, validate: isNonBlankString, normalize: trim}
    },
    API_CreateTable: {
        "tname": {defaultValue: null, validate: isNonBlankString, normalize: trim},
        "pnoun": {defaultValue: null, validate: isNonBlankString, normalize: trim}
    },
    API_AddField: {
        "label": {defaultValue: null, validate: isNonBlankString, normalize: trim},
        "type": {defaultValue: null, validate: isNonBlankString, normalize: trim},
        "mode": {defaultValue: null, validate: isNonBlankString, normalize: trim}
    },
    API_Authenticate: {
        "hour": {defaultValue: null, validate: isIntegerString, normalize: parseIntString},
        "udata": {defaultValue: null, validate: isNonBlankString, normalize: trim}
    }
};
